### PR TITLE
[FSDP2] Add gradient accumulation finalization API

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #196727
* #196726
* #196963
* #196725
* #196724
* __->__ #196640
* #196641

Add a public FSDPModule method, `finalize_gradient_accumulation`, that completes deferred gradient reduction on the caller thread. It reshards parameters and reduces the accumulated gradients into sharded `.grad`.

The autograd final callback runs on the autograd thread, so its stream waits cannot be captured in a CUDA graph. Reuse its cleanup on the caller thread, but track deferred reduction explicitly so finalization does not rerun post-backward or reduce zero gradients after a synchronized backward. Temporarily override gradient synchronization, all-reduce, reshard, and last-backward settings, then restore them.

A separate callback would duplicate iteration cleanup. The deferred-work flag keeps one cleanup path while gating the reduction. HSDP all-reduce-only deferral requires a separate collective and gradient view-out path, so reject it explicitly instead of silently skipping gradients.

Test Plan:

```bash
python3 test/distributed/_composable/fsdp/test_fully_shard_training.py TestFullyShardGradientAccumulation.test_finalize_gradient_accumulation
python3 test/distributed/_composable/fsdp/test_fully_shard_training.py TestFullyShardGradientAccumulation.test_finalize_gradient_accumulation_rejects_partial_all_reduce
python3 test/distributed/_composable/fsdp/test_fully_shard_training.py TestFullyShardCudaGraph.test_finalize_gradient_accumulation_cudagraph
UV_NO_CONFIG=1 lintrunner -a
```

AI assistance: This commit was authored with Codex.

### CODE REVIEW

File: torch/distributed/fsdp/_fully_shard/_fully_shard.py [hunk 1]
```diff
@@ -410,6 +410,68 @@ def set_is_last_backward(self, is_last_backward: bool) -> None:
...
+        is_last_backward = state._state_ctx.is_last_backward
+        try:
+            self.set_requires_gradient_sync(True)
+            self.set_reshard_after_backward(True)
+            self.set_is_last_backward(True)
+            state._root_post_backward_final_callback(
+                finalize_gradient_accumulation=True
+            )
+            state._join_comm_streams()
+        finally:
```
Comment (weifengpy) [inline torch/distributed/fsdp/_fully_shard/_fully_shard.py]: what kind of error are you expecting here? it's rare to see try-catch in fsdp2

File: test/distributed/_composable/fsdp/test_fully_shard_training.py [hunk 2]
```diff
@@ -1709,6 +1709,79 @@ def world_size(self) -> int:
...
+        ref_model = copy.deepcopy(model)
+        fully_shard(model[0], reshard_after_forward=False)
+        fully_shard(model[1], reshard_after_forward=False)
+        fully_shard(model, reshard_after_forward=False)
+
+        torch.manual_seed(42 + self.rank)
+        inputs = [torch.randn(4, 8, device=device_type.type) for _ in range(3)]
+        model.set_is_last_backward(False)
+        model.set_reshard_after_backward(False)
+        model.set_reduce_scatter_unused_params(True)
```
Comment (weifengpy) [inline test/distributed/_composable/fsdp/test_fully_shard_training.py]: is it necessary to call set_reduce_scatter_unused_params  ?

File: test/distributed/_composable/fsdp/test_fully_shard_training.py [hunk 3]
```diff
@@ -1709,6 +1709,79 @@ def world_size(self) -> int:
...
+        # Defer all gradient reduction until finalization.
+        with CommDebugMode() as comm_mode:
+            model.set_requires_gradient_sync(False)
+            for input_tensor in inputs:
+                model(input_tensor).sum().backward()
+            model.finalize_gradient_accumulation()
```
Comment (weifengpy) [inline test/distributed/_composable/fsdp/test_fully_shard_training.py]: i mostly care about UX, aligned on motivation

```
# new, calls set_is_last_backward internally, as set_is_last_backward is a bad naming now
model.set_manual_backward_finalization(True) 

model.set_requires_gradient_sync(False)
model.set_reshard_after_backward(False)
for batch in microbatches:
    model(batch).sum().backward()
model.set_requires_gradient_sync(True) # be explicit about on/off in training loop
model.set_reshard_after_backward(True)

# this method is useful with/without grad accumulation
model.finalize_backward()
```

this way UX becomes quite consistent across FSDP2 and DDP: https://github.com/pytorch/pytorch/pull/196651

File: torch/distributed/fsdp/_fully_shard/_fsdp_state.py [hunk 4]
```diff
@@ -437,6 +450,25 @@ def _root_post_backward_final_callback(self) -> None:
...
+        # then join it back to the current stream.
+        current_stream_event = self._device_handle.Event()
+        current_stream_event.record(current_stream)
+        for stream in (
+            self._comm_ctx.all_gather_copy_in_stream,
+            self._comm_ctx.all_gather_stream,
+            self._comm_ctx.reduce_scatter_stream,
+            self._comm_ctx.all_reduce_stream,
+        ):
+            stream.wait_event(current_stream_event)
```
Comment (SherlockNoMad) [inline torch/distributed/fsdp/_fully_shard/_fsdp_state.py:467]: I am not sure if this wait is actually needed...

File: torch/distributed/fsdp/_fully_shard/_fsdp_state.py [hunk 5]
```diff
@@ -437,6 +450,25 @@ def _root_post_backward_final_callback(self) -> None:
                 self._comm_ctx.reduce_scatter_states.clear()
             self._state_ctx.post_backward_final_callback_queued = False
 
+    def _join_comm_streams(self) -> None:
```
Comment (SherlockNoMad) [inline torch/distributed/fsdp/_fully_shard/_fsdp_state.py:453]: IIUC, finalize_backward() is called after backward() and before optimizer.step()? 

with _join_comm_streams(), this is enforcing all  reduce_scatter communication to complete before optimizer starts. 

If latter we want to overlap optimizer compute with reduce_scatter, this would need to be relaxed. 

Not a blocking comment tho... just sharing observation.

**Review by SherlockNoMad — APPROVED** [review]
@weifengpy should do the major signoff. 
not fsdp2 expert, but changes looks ok to me

File: torch/distributed/fsdp/_fully_shard/_fsdp_state.py (hunk 5 above)
Comment (weifengpy) [reply torch/distributed/fsdp/_fully_shard/_fsdp_state.py:453]: when grad clipping norm is not required, we directly go with optizier in the backward code path (opt.step as backward hooks). otherwise we need to express layer by layer backward finalize

**Review by sanketpurandare — APPROVED** [review]

File: torch/distributed/fsdp/_fully_shard/_fsdp_state.py (hunk 4 above)
Comment (weifengpy) [reply torch/distributed/fsdp/_fully_shard/_fsdp_state.py:467]: i will go far and say is this `_join_comm_streams` needed?

```
when is_last_backward=True. _root_post_backward_final_callback():
  - Calls FSDPParamGroup.finalize_backward(), which waits on the latest post-reduction events and any
    outstanding per-group all-gather.
  - Directly waits on retained reduce-scatter completion events before clearing their input buffers.

HEAD’s new finalize_backward() sets is_last_backward=True and invokes that callback. Therefore, those waits already execute before _join_comm_streams() is called.
```

File: torch/distributed/fsdp/_fully_shard/_fsdp_state.py [hunk 6]
```diff
@@ -415,7 +417,18 @@ def _root_post_backward_final_callback(self) -> None:
...
-                    if fsdp_param_group._training_state != TrainingState.POST_BACKWARD:
+                    if manual_finalization:
+                        if fsdp_param_group._deferred_gradient_reduction:
+                            # set_requires_gradient_sync(False) deferred this
+                            # parameter group's reduction.
+                            fsdp_param_group.post_backward()
+                        else:
+                            # This group already reduced or did not participate
+                            # in backward.
+                            fsdp_param_group.reshard()
```
Comment (weifengpy) [inline torch/distributed/fsdp/_fully_shard/_fsdp_state.py:428]: bug: reshards anyway when reshard_after_backward=False

File: torch/distributed/fsdp/_fully_shard/_fsdp_state.py [hunk 7]
```diff
@@ -415,7 +417,18 @@ def _root_post_backward_final_callback(self) -> None:
                 # autograd backward order and preserving RS overlap for
                 # per-param-mesh modules whose inputs lack gradients.
                 for fsdp_param_group in reversed(state._fsdp_param_groups):
-                    if fsdp_param_group._training_state != TrainingState.POST_BACKWARD:
+                    if manual_finalization:
```
Comment (weifengpy) [inline torch/distributed/fsdp/_fully_shard/_fsdp_state.py:420]: i was verifying if this new manual_finalization argument are truly needed and seems we can avoid it and reuse existing _root_post_backward_final_callback directly, and naturally avoid  bug on resharding when reshard_after_backward=False

```
diff --git a/torch/distributed/fsdp/_fully_shard/_fsdp_state.py b/torch/distributed/fsdp/_fully_shard/_fsdp_state.py
index 4bf0c3b6108..076b0f49116 100644
--- a/torch/distributed/fsdp/_fully_shard/_fsdp_state.py
+++ b/torch/distributed/fsdp/_fully_shard/_fsdp_state.py
@@ -398,9 +398,7 @@ class FSDPState(_State):
             return grad
 
     @_dynamo_disable
-    def _root_post_backward_final_callback(
-        self, manual_finalization: bool = False
-    ) -> None:
+    def _root_post_backward_final_callback(self) -> None:
         logger.debug("FSDP::root_post_backward")
         with torch.profiler.record_function("FSDP::root_post_backward_callback"):
             # Reset per-iteration state. With chunked loss, each standalone
@@ -417,57 +415,30 @@ class FSDPState(_State):
                 # autograd backward order and preserving RS overlap for
                 # per-param-mesh modules whose inputs lack gradients.
                 for fsdp_param_group in reversed(state._fsdp_param_groups):
-                    if manual_finalization:
-                        if fsdp_param_group._deferred_gradient_reduction:
-                            # set_requires_gradient_sync(False) deferred this
-                            # parameter group's reduction.
-                            fsdp_param_group.post_backward()
-                        else:
-                            # This group already reduced or did not participate
-                            # in backward.
-                            fsdp_param_group.reshard()
-                    elif (
-                        fsdp_param_group._training_state != TrainingState.POST_BACKWARD
-                    ):
+                    if fsdp_param_group._training_state != TrainingState.POST_BACKWARD:
                         # Run post-backward in case forward inputs did not require
                         # gradient so the autograd backward did not run
                         fsdp_param_group.post_backward()
                     fsdp_param_group._training_state = TrainingState.IDLE
                 state._training_state = TrainingState.IDLE
-                if self._state_ctx.is_last_backward:
-                    for fsdp_param_group in state._fsdp_param_groups:
-                        fsdp_param_group.finalize_backward()
             if self._state_ctx.is_last_backward:
-                self._comm_ctx.post_forward_order.clear()
-                # Wait on and release any retained reduce-scatter input buffers:
-                # the last module's (which no later module's rs_wait clears) and,
-                # when set_reduce_scatter_max_input_buffers retains more than
-                # one in flight, the rest. The compute stream (which reuses the
-                # memory) is ordered past each reduce-scatter first.
-                for rs_state in self._comm_ctx.reduce_scatter_states:
-                    if rs_state.event is not None:
-                        self._device_handle.current_stream().wait_event(rs_state.event)
-                self._comm_ctx.reduce_scatter_states.clear()
+                self._finalize_backward()
             self._state_ctx.post_backward_final_callback_queued = False
 
-    def _join_comm_streams(self) -> None:
-        if self._device.type == "cpu":
-            return
-        current_stream = self._device_handle.current_stream()
-        # Connect each communication stream to the current CUDA graph capture,
-        # then join it back to the current stream.
-        current_stream_event = self._device_handle.Event()
-        current_stream_event.record(current_stream)
-        for stream in (
-            self._comm_ctx.all_gather_copy_in_stream,
-            self._comm_ctx.all_gather_stream,
-            self._comm_ctx.reduce_scatter_stream,
-            self._comm_ctx.all_reduce_stream,
-        ):
-            stream.wait_event(current_stream_event)
-            join_event = self._device_handle.Event()
-            join_event.record(stream)
-            current_stream.wait_event(join_event)
+    def _finalize_backward(self) -> None:
+        for state in self._state_ctx.all_states:
+            for fsdp_param_group in state._fsdp_param_groups:
+                fsdp_param_group.finalize_backward()
+        self._comm_ctx.post_forward_order.clear()
+        # Wait on and release any retained reduce-scatter input buffers:
+        # the last module's (which no later module's rs_wait clears) and,
+        # when set_reduce_scatter_max_input_buffers retains more than
+        # one in flight, the rest. The compute stream (which reuses the
+        # memory) is ordered past each reduce-scatter first.
+        for rs_state in self._comm_ctx.reduce_scatter_states:
+            if rs_state.event is not None:
+                self._device_handle.current_stream().wait_event(rs_state.event)
+        self._comm_ctx.reduce_scatter_states.clear()
 
     def _register_pre_backward_hook(self, output: Any) -> Any:
         if not torch.is_grad_enabled():
```

Comment (weifengpy) [reply torch/distributed/fsdp/_fully_shard/_fsdp_state.py:420]: then _fully_shard can reuse `_finalize_backward()` code

```
  -        self.set_is_last_backward(True)
  -        state._root_post_backward_final_callback(manual_finalization=True)
  -        state._join_comm_streams()
  -        self.set_is_last_backward(False)
  +        for group in param_groups:
  +            if group._deferred_gradient_reduction:
  +                group.post_backward()
  +            elif group.reshard_after_backward:
  +                group.reshard()
  +            group._training_state = TrainingState.IDLE
  +        state._finalize_backward()
```

**Review by weifengpy — COMMENTED** [review]
done with review
* fix bug for reshard_after_backward=True case
* reuse `_root_post_backward_final_callback` internal code without introducing manual_finalization argument

**Review by weifengpy — APPROVED** [review]
